common/namer/namer.go

```diff
@@ -1,57 +1,131 @@
 package namer
 
 import (
+	"fmt"
+	"hash/fnv"
+	"io"
+	"math"
 	"strings"
 
 	"github.com/pulumi/pulumi/sdk/v3/go/pulumi"
+	"github.com/samber/lo"
 )
 
 const nameSep = "-"
 
+// In case of truncation, size of the hash suffix in percentage of the full size
+const hashSizePercent = 10
+
+// Used only for tests to skip hash suffix to ease test results validation
+var noHash = false
+
 type Namer struct {
-	ctx    *pulumi.Context
-	prefix string
+	ctx      *pulumi.Context
+	prefixes []string
 }
 
 func NewNamer(ctx *pulumi.Context, prefix string) Namer {
+	if prefix == "" {
+		return Namer{
+			ctx:      ctx,
+			prefixes: []string{},
+		}
+	}
 	return Namer{
-		ctx:    ctx,
-		prefix: prefix,
+		ctx:      ctx,
+		prefixes: []string{prefix},
 	}
 }
 
 func (n Namer) WithPrefix(prefix string) Namer {
-	childNamer := Namer{}
-	if n.prefix != "" {
-		childNamer.prefix = n.prefix + nameSep
+	return Namer{
+		ctx:      n.ctx,
+		prefixes: append(n.prefixes, prefix),
 	}
-	childNamer.prefix += prefix
-	return childNamer
 }
 
 func (n Namer) ResourceName(parts ...string) string {
 	if len(parts) == 0 {
 		panic("Resource name requires at least one part to generate name")
 	}
 
-	var resourceName string
-	if n.prefix != "" {
-		resourceName += n.prefix + nameSep
-	}
-
-	return resourceName + strings.Join(parts, nameSep)
+	return joinWithMaxLength(math.MaxInt, append(n.prefixes, parts...))
 }
 
-func (n Namer) DisplayName(parts ...pulumi.StringInput) pulumi.StringInput {
+func (n Namer) DisplayName(maxLen int, parts ...pulumi.StringInput) pulumi.StringInput {
 	var convertedParts []interface{}
 	for _, part := range parts {
 		convertedParts = append(convertedParts, part)
 	}
 	return pulumi.All(convertedParts...).ApplyT(func(args []interface{}) string {
-		strArgs := []string{n.ctx.Stack()}
+		strArgs := make([]string, 1, 1+len(n.prefixes)+len(args))
+		strArgs[0] = n.ctx.Stack()
+		strArgs = append(strArgs, n.prefixes...)
```

Comment:
<!-- Thoughts represent an idea that popped up from reviewing. These comments are non-blocking by nature, but they are extremely valuable and can lead to more focused initiatives and mentoring opportunities. -->
**💭 thought**
Seen it here it is weird to prepend everything with stack name when we call namer parts "prefixes". We might come back to the naming - not related to this PR

Reply:
Indeed, the “prefixes” are before the “parts…”, but the stack name is at the beginning of everything.